src/USER-MISC/fix_charge_regulation.cpp

```diff
@@ -0,0 +1,1335 @@
+        random_equal(NULL), random_unequal(NULL),
+        idftemp(NULL), ptype_ID(NULL) {
+
+  // Region restrictions not yet implemented ..
+
+  vector_flag = 1;
+  size_vector = 8;
+  global_freq = 1;
+  extvector = 0;
+  restart_global = 1;
+  time_depend = 1;
+  cr_nmax = 0;
+  overlap_flag = 0;
+  energy_stored = 0;
+
+  // necessary to specify the free ion types
+  cation_type = utils::inumeric(FLERR, arg[3], false, lmp);
+  anion_type = utils::inumeric(FLERR, arg[4], false, lmp);
+
+  // set defaults and read optional arguments
+  options(narg - 5, &arg[5]);
+
+  if (nevery <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (nmc < 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (llength_unit_in_nm < 0.0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (*target_temperature_tcp < 0.0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (seed <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (cation_type <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (anion_type <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (reaction_distance < 0.0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (salt_charge[0] <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if (salt_charge[1] >= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  if ((salt_charge[1] % salt_charge[0] != 0) && (salt_charge[0] % salt_charge[1] != 0))
+    error->all(FLERR,
+               "Illegal fix charge_regulation command, multivalent cation/anion charges are allowed, "
+               "but must be divisible, e.g. (3,-1) is fine, but (3,-2) is not implemented");
+
+  if (pmcmoves[0] < 0 || pmcmoves[1] < 0 || pmcmoves[2] < 0)
+    error->all(FLERR, "Illegal fix charge_regulation command");
+  if (acid_type < 0) pmcmoves[0] = 0;
+  if (base_type < 0) pmcmoves[1] = 0;
+  // normalize
+  double psum = pmcmoves[0] + pmcmoves[1] + pmcmoves[2];
+  if (psum <= 0) error->all(FLERR, "Illegal fix charge_regulation command");
+  pmcmoves[0] /= psum;
+  pmcmoves[1] /= psum;
+  pmcmoves[2] /= psum;
+
+  force_reneighbor = 1;
+  next_reneighbor = update->ntimestep + 1;
+  random_equal = new RanPark(lmp, seed);
+  random_unequal = new RanPark(lmp, seed);
+  nacid_attempts = 0;
+  nacid_successes = 0;
+  nbase_attempts = 0;
+  nbase_successes = 0;
+  nsalt_attempts = 0;
+  nsalt_successes = 0;
+}
+
+Fix_charge_regulation::~Fix_charge_regulation() {
+
+  memory->destroy(ptype_ID);
+
+  delete random_equal;
+  delete random_unequal;
+
+  if (group) {
+    int igroupall = group->find("all");
+    neighbor->exclusion_group_group_delete(exclusion_group, igroupall);
+  }
+}
+
+int Fix_charge_regulation::setmask() {
+  int mask = 0;
+  mask |= PRE_EXCHANGE;
+  return mask;
+}
+
+void Fix_charge_regulation::init() {
+
+  triclinic = domain->triclinic;
+
+  char *id_pe = (char *) "thermo_pe";
```

Comment:
This line is not needed since `Modify::find_compute()` will (now) accept constant strings, so you can use the constant directly in the call in the next line.